core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -254,31 +272,65 @@ class TransactionMarkerChannelManager(config: KafkaConfig,
                                  result: TransactionResult, coordinatorEpoch: Int,
                                  topicPartitions: immutable.Set[TopicPartition]): Unit = {
     val txnTopicPartition = txnStateManager.partitionFor(transactionalId)
-    val partitionsByDestination: immutable.Map[Node, immutable.Set[TopicPartition]] = topicPartitions.groupBy { topicPartition: TopicPartition =>
-      var brokerNode: Option[Node] = None
+    val partitionsByDestination: immutable.Map[Option[Node], immutable.Set[TopicPartition]] = topicPartitions.groupBy { topicPartition: TopicPartition =>
```

Comment:
I agree it is a bit odd, though with the above path we need to traverse the map twice with `metadataCache.getPartitionLeaderEndpoint`, one for `partition` and one for `groupBy`. Since this code is part of a sorta critical path I'm not sure which one is more efficient?